logstash-core/src/main/java/org/logstash/ackedqueue/Queue.java

```diff
@@ -114,67 +130,91 @@ public void open() throws IOException {
 
-            // track the seqNum as we add this new tail page, prevent empty tailPage with a minSeqNum of 0 to reset seqNum
-            if (tailPage.maxSeqNum() > this.seqNum) { this.seqNum = tailPage.maxSeqNum(); }
+            // but checkpoint it to update the firstUnackedPageNum if it changed
+            this.headPage.checkpoint();
+        } else {
+            // head page is non-empty, transform it into a tail page and create a new empty head page
 
-            insertTailPage(tailPage);
+            PageIO pageIO = this.pageIOFactory.build(headCheckpoint.getPageNum(), this.pageCapacity, this.dirPath);
+            TailPage p = new TailPage(headCheckpoint, this, pageIO);
+            p.checkpoint();
+            add(headCheckpoint, pageIO);
 
             headPageNum = headCheckpoint.getPageNum() + 1;
             newCheckpointedHeadpage(headPageNum);
+
+            // track the seqNum as we add this new tail page, prevent empty tailPage with a minSeqNum of 0 to reset seqNum
+            if (headCheckpoint.maxSeqNum() > this.seqNum) { this.seqNum = headCheckpoint.maxSeqNum(); }
+         }
+
+
+        // only activate the first tail page
+        if (tailPages.size() > 0) {
+            this.tailPages.get(0).getPageIO().activate();
         }
 
         // TODO: here do directory traversal and cleanup lingering pages? could be a background operations to not delay queue start?
 
         this.closed.set(false);
     }
 
-    // insert a recovered tail page into the tail pages state tracking
-    // and purge it if it is found to be fully acked
-    private void insertTailPage(TailPage p) throws IOException {
-        if (!p.isFullyAcked()) {
-            this.tailPages.add(p);
-            if (!p.isFullyRead()) {
-                this.unreadTailPages.add(p);
-                this.unreadCount += p.unreadCount();
+    private void add(Checkpoint checkpoint, PageIO pageIO) throws IOException {
+        if (checkpoint.isFullyAcked()) {
+            // first make sure any fully acked page per the checkpoint is purged
+            try {
+                pageIO.purge();
+            } catch (NoSuchFileException e) {
+                // ignore
+            }
+
+            if (this.tailPages.size() == 0) {
+                // this is the first tail page and it is fully acked so just purge it
+                this.checkpointIO.purge(this.checkpointIO.tailFileName(checkpoint.getPageNum()));
+            } else {
+                // create a tail page with a null PageIO and add it to tail pages but not unreadTailPages
+                // since it is fully read since fully acked
+                this.tailPages.add(new TailPage(checkpoint, this, null));
             }
         } else {
-            // for some reason we found a fully acked page, let's purge it.
-            p.purge();
+            TailPage p = new TailPage(checkpoint, this, pageIO);
+            this.tailPages.add(p);
+            this.unreadTailPages.add(p);
```

Comment:
For @jsvd and @colinsurprenant - is this where Joao would increment the on-disk bytesize of the queue?